### PR TITLE
API: Adds an api to get user information

### PR DESCRIPTION
**Pull request by pratap0007**
This api returns github information of a logged in user
endpoint: `/user/info`

Signed-off-by: Shiv Verma <shverma@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/design/types/type.go [hunk 1]
```diff
@@ -368,3 +368,17 @@ var RefreshToken = Type("RefreshToken", func() {
 	Description("Refresh Token for User")
 	Attribute("refresh", Token, "Refresh Token for user")
 })
+
+var UserData = Type("UserData", func() {
+	Description("Github user Information")
+	Attribute("id", UInt, "Github id for User", func() {
+		Example("id", 0)
+	})
+	Attribute("name", String, "Github user name", func() {
```
Comment (SM43) [inline api/design/types/type.go:377]: we will need another field for Github ID

File: api/design/user.go [hunk 2]
```diff
@@ -83,4 +83,32 @@ var _ = Service("user", func() {
...
+		Payload(func() {
+			Token("accessToken", String, "Access Token of User", func() {
+				Example("accessToken", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9."+
+					"eyJleHAiOjE1Nzc4ODM2MDAsImlhdCI6MTU3Nzg4MDAwMCwiaWQiOjExLCJpc3MiOiJUZWt0b24gSHViIiwic2NvcGVzIjpbInJlZnJlc2g6dG9rZW4iXSwidHlwZSI6InJlZnJlc2gtdG9rZW4ifQ."+
+					"4RdUk5ttHdDiymurlZ_f7Uy5Pas3Lq9w04BjKQKRiCE")
+			})
+			Required("accessToken")
+		})
+		Result(func() {
+			Attribute("data", types.UserData, "Github user Inoformation")
```
Comment (SM43) [inline api/design/user.go:100]: Github user Inoformation -> User Information

File: api/pkg/service/auth/auth.go [hunk 3]
```diff
@@ -138,6 +139,7 @@ func (r *request) addUser(ghUser *github.User) (*model.User, error) {
 	if user.GithubName == "" {
 		user.GithubName = ghUser.GetName()
 		user.Type = model.NormalUserType
+		user.AvatarURL = ghUser.GetAvatarURL()
```
Comment (SM43) [inline api/pkg/service/auth/auth.go:143]: we will need another check here for existing users
```
if user.AvatarURL == "" {
		user.AvatarURL = ghUser.GetAvatarURL()
}
```

File: api/pkg/testutils/utils.go [hunk 4]
```diff
@@ -137,3 +137,25 @@ func (tc *TestConfig) AddScopesForUser(userID uint, scopes []string) error {
 	}
 	return nil
 }
+
+
+// AccessTokenForUser returns access JWT for user with access:token scope
+// User will have same github login and github name in db
+func (tc *TestConfig) AccessTokenForUser(name string) (*model.User, string, error) {
```
Comment (SM43) [inline api/pkg/testutils/utils.go:144]: why do we need this?

Comment (SM43) [reply api/pkg/testutils/utils.go:144]: `UserWithScopes` is already there

File: api/pkg/db/model/model.go [hunk 5]
```diff
@@ -1,128 +1,129 @@
...
+		ResourceID uint
+		TagID      uint
+	}
+
+	User struct {
+		gorm.Model
+		AgentName            string
+		GithubLogin          string
+		GithubName           string
+		AvatarURL            string
```
Comment (piyush-garg) [inline api/pkg/db/model/model.go:92]: Where it will be used?

File: api/pkg/service/auth/auth.go [hunk 6]
```diff
@@ -139,6 +140,10 @@ func (r *request) addUser(ghUser *github.User) (*model.User, error) {
 		user.GithubName = ghUser.GetName()
 		user.Type = model.NormalUserType
 
+		if user.AvatarURL == "" {
+			user.AvatarURL = ghUser.GetAvatarURL()
+	}
```
Comment (piyush-garg) [inline api/pkg/service/auth/auth.go:145]: Needs indentation

File: api/pkg/db/model/model.go (hunk 5 above)
Comment (pratap0007) [reply api/pkg/db/model/model.go:92]: while fetching user details after authentication

File: api/design/user.go [hunk 7]
```diff
@@ -83,4 +83,32 @@ var _ = Service("user", func() {
 			Response("invalid-scopes", StatusForbidden)
 		})
 	})
+
+	Method("Info", func() {
+		Description("Get the user Info")
+		Security(types.JWTAuth, func() {
```
Comment (SM43) [inline api/design/user.go:89]: Security(types.JWTAuth)

File: api/pkg/service/auth/auth.go [hunk 8]
```diff
@@ -139,6 +140,10 @@ func (r *request) addUser(ghUser *github.User) (*model.User, error) {
 		user.GithubName = ghUser.GetName()
 		user.Type = model.NormalUserType
 
+		if user.AvatarURL == "" {
```
Comment (SM43) [inline api/pkg/service/auth/auth.go:143]: this should be outside of `if user.GithubName == "" {}`

Comment (SM43) [reply api/pkg/service/auth/auth.go:143]:
```
        // User already exist, check if GitHub Name is empty
	// If Name is empty, then user is inserted through config.yaml
	// Update user with remaining details

	if user.GithubName == "" {
		user.GithubName = ghUser.GetName()
		user.Type = model.NormalUserType
                user.AvatarURL = ghUser.GetAvatarURL()

		if err = r.db.Save(&user).Error; err != nil {
			r.log.Error(err)
			return nil, err
		}
	}
	// For existing user, check if URL is not added
        if user.AvatarURL == "" {
		user.AvatarURL = ghUser.GetAvatarURL()
		if err = r.db.Save(&user).Error; err != nil {
		      r.log.Error(err)
		      return nil, err
		  }
	}


```

Comment (SM43) [reply api/pkg/service/auth/auth.go:143]: also, could you please add a test for this.. where user in DB doesn't have `URL` but after login URL is added

File: api/pkg/service/user/user.go [hunk 9]
```diff
@@ -181,3 +181,20 @@ func (r *request) newRefreshToken() (*user.NewRefreshTokenResult, error) {
...
+
+	data, err := s.User(ctx)
+	if err != nil {
+		return nil, err
+	}
+	res = &user.InfoResult{ Data: &user.UserData{
+		GithubID: data.GithubLogin,
+		Name: data.GithubName,
+		AvatarURL: data.AvatarURL,
+	},
```
Comment (SM43) [inline api/pkg/service/user/user.go:197]: Could you please format this code

File: api/pkg/service/user/user.go [hunk 10]
```diff
@@ -181,3 +181,20 @@ func (r *request) newRefreshToken() (*user.NewRefreshTokenResult, error) {
 
 	return &user.NewRefreshTokenResult{Data: data}, nil
 }
+
+
+// Get the user Info
+func (s *service) Info(ctx context.Context, p *user.InfoPayload) (res *user.InfoResult, err error) {
```
Comment (SM43) [inline api/pkg/service/user/user.go:187]: `(*user.InfoResult, error)` consistent for all functions 😄

File: api/design/user.go [hunk 11]
```diff
@@ -83,4 +83,31 @@ var _ = Service("user", func() {
...
+		Payload(func() {
+			Token("accessToken", String, "Access Token of User", func() {
+				Example("accessToken", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9."+
+					"eyJleHAiOjE1Nzc4ODM2MDAsImlhdCI6MTU3Nzg4MDAwMCwiaWQiOjExLCJpc3MiOiJUZWt0b24gSHViIiwic2NvcGVzIjpbInJlZnJlc2g6dG9rZW4iXSwidHlwZSI6InJlZnJlc2gtdG9rZW4ifQ."+
+					"4RdUk5ttHdDiymurlZ_f7Uy5Pas3Lq9w04BjKQKRiCE")
+			})
+			Required("accessToken")
+		})
+		Result(func() {
+			Attribute("data", types.UserData, "User Inoformation")
```
Comment (sm43) [inline api/design/user.go:99]: Inoformation > Information

**Review by sm43 — CHANGES_REQUESTED** [review]

**Review by PuneetPunamiya — COMMENTED** [review]
Could you please run the `goa gen` command once since files are getting updated

File: api/go.mod [hunk 12]
```diff
@@ -27,6 +27,7 @@ require (
 	gopkg.in/h2non/gock.v1 v1.0.15
 	gorm.io/driver/postgres v1.0.2
 	gorm.io/gorm v1.20.7
+	gotest.tools v2.2.0+incompatible
```
Comment (sm43) [inline api/go.mod:30]: Could you please run `go mod tidy`

Comment (pratap0007) [reply api/go.mod:30]: I have run it already

**Pull request merged**